[Refactor] 유지보수 게시글 조회 기능 리팩토링

ghrltjdtprbs opened the pull request:
# ⭐️ [Refactor] 유지보수 게시글 조회 기능 리팩토링

- 최신순 정렬을 따로 하는 대신 데이터베이스 조회 시에 id 를 desc로 불러와 자동 정렬되게 수정
- facade 패턴 적용
- GET /repair/articles/list 조회시에 현재 로그인한 사용자에 따라 응답값 분리
- 24시간 이내에 생성된 게시물임을 나타내는 isNewArticle 로직 삭제(프론트가 처리하도록 변경)

## 🛠️ 변경 사항

커밋 메시지에 자세히 적어두었으니 커밋 메시지 참고하시길 바랍니다.

## 💡 관련 이슈

- #147

## ❗️ 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트


## ⏱️ 작업 기간

- 작업 시작일: (2024-06-03)
- 작업 종료일: (2024-06-03)

## 📌 유의사항

- 테스트 코드 수정은 후에 커밋 예정이니 수정된 로직만 봐주시면 됩니다!

ghrltjdtprbs committed:
[refactor] 최신순 정렬을 조회시 id desc로 대체

문제 사항 :
 - 최신순으로 따로 정렬을 하는 것보다 애초에 불러올 때 id 기준으로 order by desc 로 불러오면 최신순 정렬 로직을 제거할 수 있음. (성능상으로도 이점)

변경 사항 :
 - findByOfficeIdAndProgressInAndCategory -> findByOfficeIdAndProgressInAndCategoryOrderByCreatedAtDesc
 - ORDER BY ra.id DESC 구문 추가

ghrltjdtprbs committed:
[feat] 관리자 전용 전체조회 response dto 추가

문제 사항 :
 - 전체 조회 기능에서 관리자와 소유주의 응답값이 서로 다름

변경 사항 :
 - ManagerRepairArticleResponseDto 추가

ghrltjdtprbs committed:
[refactor] 소유주 전용 전체조회 response dto 클래스명 변경

문제 사항 :
 - ManagerRepairArticleResponseDto를 추가하였으므로,  기존 RepairArticleResponseDto가 소유주의 응답값이라는 것을 더 명확하게 설정해야함

변경 사항 :
 - RepairArticleResponseDto -> OwnerRepairArticleResponseDto 이름 변경

ghrltjdtprbs committed:
[refactor] 관리자, 소유주 전체조회 분리

문제 사항 :
 - 관리자와 소유주의 게시글 전체조회 로직을 분리
 - 응답값, 정렬 기준이 틀리므로 분리하는 것이 더 합리적이라 판단

변경 사항 :
 - 기존 전체조회 로직을 OwnerArticleListService, ManagerArticleListService로 분리하여 각각의 요구사항에 맞게 커스텀

ghrltjdtprbs committed:
[refactor] 조회 관련 서비스에 facade 패턴 적용

문제 사항 :
 - 기존 전체 조회 로직을 분리하면서 facade 패턴을 적용하여 더 직관적이고 효육적인 코드 설계

변경 사항 :
 - GetRepairArticleFacade 생성
 - 매니저 여부에 따라 전체조회 로직 분리 호출
 - 그 외 조회 로직들도 facade 에 포함시킴

ghrltjdtprbs committed:
[refactor] OwnerRepairArticleController 클래스명 변경

문제 사항 :
 - 기존 기능을 manager 도 함께 이용하게 되면서 OwnerRepairArticleController 클래스명은 적절하지 않음

변경 사항 :
 - OwnerRepairArticleController -> GetRepairArticleController

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/controller/OwnerRepairArticleController.java:
```diff
@@ -26,26 +25,26 @@
 public class OwnerRepairArticleController {
 
-    private final OwnerRepairArticleService ownerRepairArticleService;
     private final AuthenticationService authenticationService;
```
파사드 패턴을 적용하셨다면 authenticationService도 파사드로 옮겨 컨트롤러의 의존성을 최대한 낮추는게 좋을것같습니다

ghrltjdtprbs replied:
앗 이부분을 놓쳤었네요! 지적 감사합니다 ㅎㅎ 수정해서 커밋했으니 확인해주세요!!👍👍

관련 커밋 : a5e232f45c10b7cc4fbf645d6612c57e3504a7ce

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/controller/OwnerRepairArticleController.java:
```diff
@@ -26,26 +25,26 @@
         UserIdAndRole userIdAndRole = authenticationService.getCurrentUserIdAndRole();
 
-        List<RepairArticleResponseDto> articles = ownerRepairArticleService.getAllRepairArticles(
+        List<?> articles = getRepairArticleFacade.getAllRepairArticles(
```
굳이 와일드 카드를 사용하신 이유가 있을까오?

ghrltjdtprbs replied:
관리자일때랑 소유주일때랑 responsedto 가 다릅니다!

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/repository/RepairArticleRepository.java:
```diff
@@ -24,8 +24,9 @@ long countByOfficeIdAndProgress(@Param("officeId") Long officeId,
-    List<RepairArticle> findByOfficeIdAndProgressInAndCategory(
+        "AND (:category IS NULL OR ra.category = :category)" +
+        "ORDER BY ra.id DESC")
+    List<RepairArticle> findByOfficeIdAndProgressInAndCategoryOrderByCreatedAtDesc(
```
쿼리에 개행이 있다면 텍스트 블록 문법을 사용하는건 어떠신가요?
"""SELECT ra FROM RepairArticle ra WHERE ra.office.id = :officeId 
        AND (:progresses IS NULL OR ra.progress IN :progresses) 
        AND (:category IS NULL OR ra.category = :category)
        ORDER BY ra.id DESC"""

asd42270 replied:
저도 이게 좋을 것 같습니다!!

ghrltjdtprbs replied:
요런게 있었네여 ㅋㅋㅋ 수정해서 커밋했으니 확인해주세요! 😊

관련 커밋 : 3f336b9dcaaa04f7621421dbfd98b4f3e9e63b91

Galmaeki reviewed (approved):
고생 많으셨슴다!
코멘트 확이 ㄴ부탁드림다!

asd42270 reviewed (approved):
고생 많으셨습니다!!

ghrltjdtprbs committed:
[fix] 테스트코드에 변경사항 반영

변경 사항 :
 - 전체조회 테스트 코드에 변경사항을 반영하였습니다.

ghrltjdtprbs committed:
[refactor] GetRepairArticleController에 AuthenticationService 제거

문제 사항 :
 - GetRepairArticleFacade에서 이미 AuthenticationService 사용
 - 컨트롤러에서 AuthenticationService를 제거 하는 것이 의존성 낮추는 것에 더 효과적

변경 사항 :
 - GetRepairArticleController에 AuthenticationService 제거
 - 관련 정보는 GetRepairArticleFacade에서 처리

ghrltjdtprbs committed:
[refactor] 매니저 전체조회시 updatedAt 정렬 제거

문제 사항 :
 - 요구 사항에 없음

변경 사항 :
 - ManagerArticleListService 에서  updatedAt 정렬 제거

ghrltjdtprbs committed:
[fix] 리팩토링 사항에 맞게 테스트 코드 수정

변경 사항 :
 - 리팩토링 사항에 맞게 테스트 코드 수정

ghrltjdtprbs committed:
[refactor] RepairArticleRepository에 텍스트 블록문법 적용

문제 사항 :
 - 피드백 반영

변경 사항 :
 - RepairArticleRepository에 텍스트 블록 문법 적용
